polar-core/src/vm.rs

```diff
@@ -1889,25 +1969,185 @@ impl PolarVirtualMachine {
+                args[1] = x;
+                self.push_goal(Goal::Query {
+                    term: term.clone_with_value(Value::Expression(Operation {
+                        operator: *op,
+                        args,
+                    })),
+                })?;
+                return Ok(QueryEvent::None);
+            } else if !handle_unbound_right_var && left.value().as_symbol().is_err() {
+                return eval(self, term);
+            }
+        }
+
+        if let Value::Variable(l) = left.value() {
+            // A variable on the left, ground on the right.
+            if let VariableState::Bound(x) = self.variable_state(l) {
+                args[0] = x;
+                self.push_goal(Goal::Query {
+                    term: term.clone_with_value(Value::Expression(Operation {
+                        operator: *op,
+                        args,
+                    })),
+                })?;
+                return Ok(QueryEvent::None);
+            } else if !handle_unbound_left_var && right.value().as_symbol().is_err() {
+                return eval(self, term);
+            }
+        }
+
+        if left.value().as_symbol().is_ok() || right.value().as_symbol().is_ok() {
+            self.add_constraint(term)?;
+            return Ok(QueryEvent::None);
+        }
+
+        eval(self, term)
+    }
+
+    /// Evaluate comparison operations.
+    fn comparison_op_helper(&mut self, term: &Term) -> PolarResult<QueryEvent> {
+        let Operation { operator: op, args } = term.value().as_expression().unwrap();
+
+        assert_eq!(args.len(), 2);
+        let left = &args[0];
+        let right = &args[1];
+
+        match (left.value(), right.value()) {
+            // TODO(ap): ExternalInstance on one side.
+            (Value::ExternalInstance(_), Value::ExternalInstance(_)) => {
+                // Generate a symbol for the external result and bind to `false` (default).
+                let (call_id, answer) =
+                    self.new_call_var("external_op_result", Value::Boolean(false));
+
+                // Check that the external result is `true` when we return.
+                self.push_goal(Goal::Unify {
+                    left: answer,
+                    right: Term::new_temporary(Value::Boolean(true)),
+                })?;
+
+                // Emit an event for the external operation.
+                Ok(QueryEvent::ExternalOp {
+                    call_id,
+                    operator: *op,
+                    args: vec![left.clone(), right.clone()],
+                })
+            }
+            _ => {
+                if !compare(*op, left, right) {
+                    self.push_goal(Goal::Backtrack)?;
+                }
+                Ok(QueryEvent::None)
+            }
+        }
+    }
+
+    // TODO(ap, dhatch): This does not strip the 3rd argument like Dot op (the unbound variable).
+    // So may be incorrect currently.
```

Comment:
This comment is a bit inscrutable. Why does not stripping the third arg lead to incorrectness?

Reply:
Because the adapter or whatever backend that handles arithmetic probably doesn't expect ternary operators. So we should rewrite `+(a, b, c)` to `c = +(a, b)` like we do for dots.

Reply:
Would you mind updating the comment to include that explanation?

Reply:
a08baa82